KAFKA-10163; Throttle Create Topic, Create Partition and Delete Topic Operations (KIP-599, Part I, Broker Changes)

dajac opened the pull request:
This PR implements the broker side changes of KIP-599, except the changes of the Rate implementation which will be addressed separately.

The PR changes/introduces the following:
- It introduces the protocol changes.
- It introduces a new quota manager `ControllerMutationQuotaManager` which is another specialization of the `ClientQuotaManager`.
- It enforces the quota in the `KafkaApis` and in the `AdminManager`. This part handles new and old clients as described in the KIP.

The PR is a good shape. The tests coverage is good. I will add few more unit tests for the api layer later one.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac commented on clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:
```diff
@@ -97,7 +97,25 @@ public static RecordingLevel forName(String name) {
+    }
 
+    public enum QuotaEnforcementType {
+        /**
+         * The quota is not enforced.
+         */
+        NONE,
+
+        /**
+         * The quota is enforced before updating the sensor. An update
+         * is rejected if the quota is already exhausted.
+         */
+        STRICT,
+
+        /**
+         * The quota is enforced after updating the sensor. An update
+         * is always accepted even if the quota is already exhausted.
+         */
+        PERMISSIVE
```
The various quota enforcement types that can be used with a `Sensor`.

dajac commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -154,32 +180,40 @@ object ClientQuotaManager {
 class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
                          private val metrics: Metrics,
                          private val quotaType: QuotaType,
+                         private val quotaEnforcementType: QuotaEnforcementType,
```
I have refactored the `ClientQuotaManager` to be a bit more generic and to accept a quota enforcement type.

rajinisivaram replied:
It looks like we don't need this change since the only one that sets a different type does its own thing.

dajac replied:
Yes, you are right. I was trying to make `ClientQuotaManager` a generic base class for all client quotas. I will still try to extract the real common parts in a base class once but that will be for a follow-up PR.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1684,25 +1685,29 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   def handleCreateTopicsRequest(request: RequestChannel.Request): Unit = {
+    // Since version 6 of the API, the quota is strictly enforced. Any topic creation
+    // above the quota is not allowed and rejected with a THROTTLING_QUOTA_EXCEEDED error.
+    val controllerMutationQuota = quotas.controllerMutation.newQuotaFor(request, 6)
```
I have introduced a new construct here in order to make the integration as lightweight as possible: `ControllerMutationQuota`. It is a small wrapper around the `Sensor` of a given user/clientId which wraps the quota enforcement and the computation of the throttle time. Three implementations are available: `StrictControllerMutationQuota` that is used for new clients, `PermissiveControllerMutationQuota` that is used for old clients, and `UnboundedControllerMutationQuota` that is used when no quota is set.

The `ControllerMutationQuota` object is then passed to the `AdminManager` to enforce the quota and later used to compute the final throttle time before sending the response. At this point, the request quota is also computed and the highest one is used.

rajinisivaram replied:
nit: `quotas.controllerMutation.newQuotaFor(request, strictSinceVersion=6)`. Then we can move the comment to the method.

dajac commented:
@junrao @rajinisivaram @apovzner Whenever you have time, I would appreciate if you could check this one out. Thanks.

dajac committed:
Protocol

dajac committed:
Sensor

dajac committed:
Quota Manager in Core

dajac committed:
cleanup

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:
```diff
@@ -97,7 +97,25 @@ public static RecordingLevel forName(String name) {
         }
+    }
 
+    public enum QuotaEnforcementType {
```
Do we really need this? I didn't find the naming of the types very intuitive. We have different variations on the usage of the APIs. But given that we have a public API that enables us to check quotas and another to record with or without the check, we could just use the existing APIs? For example, we already have record without checking quota, we have record with check quota, we have record followed by unrecord. And now we want to check without recording. It seems simpler to leave it to the quota manager to record zero with  checkQuotas first and then record the value without checkQuotas if the first one succeeds.

dajac replied:
I think that we do. I actually did exactly what you suggest in the first place but then realised that it introduces a race condition. For instance, two requests processed in two separate threads can verify the quota at the same time. If it is below the authorized quota, they would go ahead, record the value, and do the creation/deletion. That violates the strict enforcement of the quota as we don't want to accept a creation/deletion if the quota is already violated. In this case, the first request must succeed while the second must fail.

The proposed approach allows to do the quota verification and the recording within the sensor lock. That basically allows us to record the value iif the quota is not already violated.

This is probably less an issue for our existing quotas because we always accept requests, even if the quota is already violated.

Regarding the naming, we may find better ones. Would you have any suggestions?

rajinisivaram replied:
Thanks for the explanation, makes sense. 

Not sure about a better name, but it seems to me that we are changing a boolean `checkQuotas=true|false` to an enum that says `checkQuotas=false|before|after`. Perhaps something that makes that clearer.

dajac replied:
On one side, we have names that describe how the quota enforcement behave, and on the other side, we have names that describe where the quota is enforced. I personally lean towards the current proposal but I an open to discuss this. My reasoning is that the current proposal is aligned with the naming in the quota manager as well. That makes the overall more consistent IMO.

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -101,12 +148,6 @@ class AdminManager(val config: KafkaConfig,
-        topic.configs.forEach { entry =>
-          configs.setProperty(entry.name, entry.value)
-        }
-        LogConfig.validate(configs)
```
Where is this validation done now?

dajac replied:
This is done in `validateTopicCreate` (last line). Actually, I found out that we were validating the `LogConfig` twice before.

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -295,34 +312,44 @@ class AdminManager(val config: KafkaConfig,
+          topic, existingAssignment, allBrokers, newPartition.count, newPartitionsAssignment)
+
+        if (validateOnly) {
+          CreatePartitionsMetadata(topic, (existingAssignment ++ assignmentForNewPartitions).keySet)
```
Shouldn't validateOnly tell you know much you would have been throttled?

dajac replied:
That's a great question. If I remember correctly, someone brought this up in the DISCUSS thread a while back.

The main issue is that we can't use the `throttleTimeMs` field to give back this information to the client because the java client would consider this as a quota violation and would throttle the connection. If we want to do this, we would need to find another way to convey this to the client. I don't think that it is worth it.

Moreover, as the validation of the request doesn't not entail any creations nor deletions, I would find strange to actually return `ThrottlingQuotaExceededException` with a `throttleTimeMs` to the client. `ThrottlingQuotaExceededException` would actually mean "the validation has passed but you would have been throttled if you would have executed that request" which is not intuitive if you care about validating the request only.

rajinisivaram commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -234,55 +266,85 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+   * @param request client request
+   * @param throttleTimeMs Duration in milliseconds for which the channel is to be muted.
+   * @param channelThrottlingCallback Callback for channel throttling
+   * @return ThrottledChannel object
```
Doesn't return anything?

dajac replied:
Indeed, good catch.

rajinisivaram commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -20,21 +20,32 @@ import java.util.concurrent.TimeUnit
+  private val maxThrottleTimeMs = TimeUnit.SECONDS.toMillis(this.config.quotaWindowSizeSeconds)
+  private val exemptMetricName = metrics.metricName("exempt-request-time",
+    QuotaType.Request.toString, "Tracking exempt-request-time utilization percentage")
+  private val exemptSensorName = "exempt-" + QuotaType.Request
```
We might as well move the name to the companion object above since it is a constant?

dajac replied:
Yes

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,205 @@
+ */
+trait ControllerMutationQuota {
+  def isExceeded: Boolean
+  def accept(permits: Double): Unit
```
nit: maybe use the same `record` terminology we use elsewhere?

dajac replied:
Yes, that makes sense. I initially went with `accept` to make it clear that it could be refused but `record` is fine as well.

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,205 @@
+ * quota is not meant to be cached forever but rather during the lifetime of processing
+ * a request.
+ */
+trait ControllerMutationQuota {
```
We should include `Request` in the names of these traits and associated classes since these are per-request.

dajac replied:
I am not entirely sold on this one. Do you suggest to use something like `ControllerMutationRequestQuota`? My intent was to replace `Request` by `Mutation` here as the quota is actually enforced by mutation within a request. We could perhaps use `ControllerRequestQuota` though. How do you see this?

rajinisivaram replied:
Sorry, `Mutation` is fine, and we dont want to say`RequestQuota` since request quota is already a thing. My concern was that we have `quotas` which apply to entities like users over a period of time. But this seems to be something that is relevant to a specific request.

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,205 @@
+                                    private val quotaSensor: Sensor) extends ControllerMutationQuota {
+
+  private var lastThrottleTimeMs = 0L
+  private var lastRecordedTimeMs = 0L
```
Are these called `lastXXX` because we can record multiple times and we are interested in the last value?

dajac replied:
That is correct. They are basically updated every time `record` (formally `accept`) is called if the quota is violated. We keep only the last one because it gives the fresher information about how long is required to bring back the average rate to the quota.

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,205 @@
+    // between the recording and now. We do this because `throttleTime` may be called
+    // long after having recorded it (e.g. when creating topics).
+    val deltaTimeMs = time.milliseconds - lastRecordedTimeMs
+    Math.max(0, lastThrottleTimeMs - deltaTimeMs).toInt
```
Not sure how this works. We calculated throtle time before creating a topic. Then spent some time using up CPU to create the topic. Are we saying that we should throttle less because the request used up time?

dajac replied:
I realized that I could have better explained this in the comment. I actually do this to deduce the time that the request spent waiting in the purgatory.

Let's take an example. A `CreateTopicsRequest` with two topics may be able to create the first topic while the second is refused due to a quota exception with, say a throttling time of 5s. Then the request goes to the purgatory waiting until the first topic is created. That could take few seconds or even longer that the actual throttle time. In the latter case, we want the let the client retries immediately. Therefore, I deduce the time between the last recorded throttle time and now to reflect that.

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,205 @@
+ * @param quotaSensor @Sensor object with a defined quota for a given user/clientId pair
+ */
+class PermissiveControllerMutationQuota(private val time: Time,
+                                        private val quotaSensor: Sensor) extends ControllerMutationQuota {
```
Looks pretty identical to `StrictControllerMutationQuota` apart from the enforcement type, can we reuse code?

dajac replied:
Let me try to extract the common bits into a base class.

dajac replied:
I have extracted the common bits into a base class. Let me know if that works for you.

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1746,21 +1751,26 @@ class KafkaApis(val requestChannel: RequestChannel,
     val createPartitionsRequest = request.body[CreatePartitionsRequest]
+    // Since version 3 of the API, the quota is strictly enforced. Any partition creation
+    // above the quota is not allowed and rejected with a THROTTLING_QUOTA_EXCEEDED error.
+    val controllerMutationQuota = quotas.controllerMutation.newQuotaFor(request, 3)
```
As before, use named argument for `strictSinceVersion` and comment not necessary here?

dajac replied:
Yeah, that is a good idea.

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1803,14 +1813,20 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleDeleteTopicsRequest(request: RequestChannel.Request): Unit = {
+    // Since version 5 of the API, the quota is strictly enforced. Any topic deletion
+    // above the quota is not allowed and rejected with a THROTTLING_QUOTA_EXCEEDED error.
+    val controllerMutationQuota = quotas.controllerMutation.newQuotaFor(request, 5)
```
Named argument for `strictSinceVersion`

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1777,7 +1787,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         responseBody
       }
-      sendResponseMaybeThrottle(request, createResponse)
+      sendResponseMaybeThrottle(controllerMutationQuota, request, createResponse, None)
```
nit: Make the last argument a default arg or use named argument since it is not clear what None is.

dajac replied:
Ack. I went with the later.

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1777,7 +1787,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         responseBody
       }
-      sendResponseMaybeThrottle(request, createResponse)
+      sendResponseMaybeThrottle(controllerMutationQuota, request, createResponse, None)
```
nit: Make the last argument a default arg or use named argument since it is not clear what None is.

rajinisivaram commented on core/src/test/scala/unit/kafka/server/ControllerMutationQuotaTest.scala:
```diff
@@ -0,0 +1,361 @@
+
+      // The implementation of the Rate has NOT been changed yet so we have to wait past
+      // the window in order to get the avg rate bellow the quota.
+      Thread.sleep(11000) // Thread.sleep(throttleTimeMs1)
```
Can we test using mock time or a different approach that doesn't need these long sleeps?

dajac replied:
I do agree that this is pretty annoying. I am a bit reluctant to using mock time here as these are real integration tests so I would like to test the real conditions. When we will have the improvement for the quota, the sleep time will be <5s and we have 3 of them in the whole test suite. IMO, 15s is not that bad. We do similar in the other quota integration tests for example.

I may be able to reduce it a bit by tweaking the quota used. Let me try.

rajinisivaram replied:
In these type of tests, it is better to use `waitUntil` some condition rather than sleep for a long time. The sleep seems very long because there are multiple tests like this. If it is possible to wait for some condition that typically finishes sooner but may some times take longer, that would be better.

dajac replied:
That makes sense. I have changed them to using `waitUntilTrue`.

rajinisivaram reviewed (commented):
@dajac Thanks for the PR, looks good, left a few comments

dajac committed:
Address review

dajac committed:
Address Anna's comment in #8977.

dajac committed:
address review

dajac commented:
@rajinisivaram Thanks for your review. I have addressed all your comments module the two regarding the naming of the type and the classes which are still under discussions.

dajac committed:
Remove QuotaEnforcementType

rajinisivaram commented:
ok to test

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,240 @@
+/**
+ * The StrictControllerMutationQuota defines a strict quota for a given user/clientId pair. The
+ * quota is strict meaning that it does not accept any mutations once the quota is exhausted until
+ * it gets back to the defined rate.
```
There are two differences between Strict and Permissive quotas:
1) As described above, it does not accept any mutations if quota is already exhausted
2) It does not throttle for any number of mutations if quota is not already exhausted. This is different from the other one which throttles if quota will be exceeded due to the current request.

Should we document the second behaviour as well? I am guessing 2) will be limited by the burst value. Where would that be checked?

dajac replied:
Good point. Let me complete the scaladoc.

Regarding the burst, there is no explicit check for it as it is implicitly enforced based on the length of the time window of the rate. The maximum burst is defined by the length of the window (# samples * sample length).

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,240 @@
+  /**
+   * Returns a ControllerMutationQuota based on `strictSinceVersion`. It returns a strict
+   * quota if the version is equal to or above of the `strictSinceVersion`, a permissive
+   * quota if the version is bellow, and a unbounded quota if the quota is disabled.
```
typo: below

dajac committed:
fix checkstyle errors

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,240 @@
+  }
+
+  /**
+   * Returns a StrictControllerMutationQuota for the given session/clientId pair or
```
nit: user/clientId pair?

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,240 @@
+    newStrictQuotaFor(request.session, request.header.clientId)
+
+  /**
+   * Returns a PermissiveControllerMutationQuota for the given session/clientId pair or
```
nit: user/clientId pair?

rajinisivaram commented on core/src/main/scala/kafka/server/ControllerMutationQuotaManager.scala:
```diff
@@ -0,0 +1,240 @@
+   * When the quota is strictly enforced. Any operation above the quota is not allowed
+   * and rejected with a THROTTLING_QUOTA_EXCEEDED error.
+   *
+   * @param request The request to extract the session and the clientId from
```
nit: user and clientId of the session

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -864,9 +868,12 @@ object KafkaConfig {
   val NumQuotaSamplesDoc = "The number of samples to retain in memory for client quotas"
   val NumReplicationQuotaSamplesDoc = "The number of samples to retain in memory for replication quotas"
   val NumAlterLogDirsReplicationQuotaSamplesDoc = "The number of samples to retain in memory for alter log dirs replication quotas"
+  val NumControllerQuotaSamplesDoc = "The number of samples to retain in memory for controller mutations replication quotas"
```
nit: remove `replication`

rajinisivaram commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -222,6 +222,10 @@ class MetadataCache(brokerId: Int) extends Logging {
   }
 
+  def numPartitions(topic: String): Int = {
+    metadataSnapshot.partitionStates.get(topic).map(_.size).getOrElse(0)
```
nit: return Option[Int] and the caller can do `getOrElse(0)`?

rajinisivaram commented on core/src/test/scala/unit/kafka/server/ControllerMutationQuotaTest.scala:
```diff
@@ -0,0 +1,352 @@
+      // Theoretically, the throttle time should be below or equal to:
+      // ((30 / 10) - 2) / 2 * 10 = 5s
+      val (throttleTimeMs1, errors1) = createTopics(TopicsWith30Partitions, StrictCreateTopicsRequestVersion)
+      assertTrue((5000 - throttleTimeMs1) < 1000)
```
We should log the throttle time in the assertion message. Not sure if `< 1000` will result in flaky tests, will see how the PR builds in Jenkins do.

dajac replied:
Good point. Regarding the flakyness, I never had a failure locally but I suppose that may happen if Jenkins is really slow.

rajinisivaram commented on core/src/test/scala/unit/kafka/server/ControllerMutationQuotaTest.scala:
```diff
@@ -0,0 +1,352 @@
+      // Theoretically, the throttle time should be below or equal to:
+      // ((60 / 10) - 2) / 2 * 10 = 20s
+      val (throttleTimeMs, errors) = createTopics(TopicsWith30Partitions, PermissiveCreateTopicsRequestVersion)
+      assertTrue((20000 - throttleTimeMs) < 1000)
```
nit: Include actual `throttleTimeMs` in assertion message (multiple places)

rajinisivaram commented on core/src/test/scala/unit/kafka/server/ThrottledChannelExpirationTest.scala:
```diff
@@ -14,10 +14,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
```
Seems to be changing a file for no reason

rajinisivaram commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -20,13 +20,11 @@ import java.util.concurrent.{ConcurrentHashMap, TimeUnit}
 import java.util.concurrent.locks.ReentrantReadWriteLock
 
 import scala.collection.Seq
-
```
Revert changes to file since nothing has actually changed?

rajinisivaram reviewed (commented):
@dajac Thanks for the updates, left a few more minor comments

dajac committed:
address comments

dajac committed:
fixup

dajac commented:
@rajinisivaram I just pushed an update which addresses your comments.

rajinisivaram reviewed (approved):
@dajac Thanks for the updates, LGTM

rajinisivaram commented:
retest this please

dajac committed:
fixup

rajinisivaram commented:
retest this please

rajinisivaram commented:
retest this please

dajac committed:
fix flaky tests

rajinisivaram commented:
retest this please

rajinisivaram commented:
retest this please

rajinisivaram commented:
ok to test

dajac committed:
Enhance tests

dajac committed:
fix flaky test

rajinisivaram commented:
ok to test

rajinisivaram commented:
retest this please

rajinisivaram commented:
ok to test

rajinisivaram commented:
Merged to trunk, test failures were unrelated.
